gpflow/kernels.py

```diff
@@ -826,6 +826,124 @@ def on_separate_dimensions(self):
             return not overlapping
 
 
+class Convolutional(Kernel):
+    """
+    Conv
+    Plain convolutional kernel.
+    """
+
+    def __init__(self, basekern, img_size, patch_size, colour_channels=1):
+        Kernel.__init__(self, np.prod(img_size))
+        self.img_size = img_size
+        self.patch_size = patch_size
+        self.basekern = basekern
+        self.colour_channels = colour_channels
+
+        if self.basekern.input_dim != np.prod(patch_size):
+            raise ValueError("Basekern input dimensions must be consistent with patch size.")
+
+    @params_as_tensors
+    def get_patches(self, X):
+        """
+        Extracts patches from the images X. Patches are extracted separately for each of the colour channels.
+        :param X: (N x input_dim)
+        :return: Patches (N, num_patches, patch_size)
+        """
+        # castX = tf.transpose(
+        #     tf.reshape(tf.cast(X, tf.float32, name="castX"), tf.stack([tf.shape(X)[0], -1, self.colour_channels])),
+        #     [0, 2, 1])
+        # castX = tf.cast(X, tf.float32, name="castX")
+
+        # Roll the colour channel to the front, so it appears to `tf.extract_image_patches()` as separate images. Then
+        # extract patches and reshape to have the first axis the same as the number of images. The separate patches will
+        # then be in the second axis.
+        castX = tf.transpose(
+            tf.reshape(X, tf.stack([tf.shape(X)[0], -1, self.colour_channels])),
+            [0, 2, 1])
+        # castX = tf.cast(castX, tf.float32, name="castX")
+        patches = tf.extract_image_patches(
+            tf.reshape(castX, [-1, self.img_size[0], self.img_size[1], 1], name="rX"),
+            [1, self.patch_size[0], self.patch_size[1], 1],
+            [1, 1, 1, 1],
+            [1, 1, 1, 1], "VALID")
+        shp = tf.shape(patches)  # img x out_rows x out_cols
+        return tf.cast(tf.reshape(patches,
+                                  [tf.shape(X)[0], self.colour_channels * shp[1] * shp[2], shp[3]]),
+                       settings.float_type)
+
+    @params_as_tensors
+    def K(self, X, X2=None):
+        Xp = self.get_patches(X)
+        Xp = tf.reshape(Xp, (-1, self.patch_len))
+        Xp2 = tf.reshape(self.get_patches(X2), (-1, self.patch_len)) if X2 is not None else None
+
+        bigK = self.basekern.K(Xp, Xp2)  # N * num_patches x N * num_patches
+        K = tf.reduce_sum(tf.reshape(bigK,
```

Comment:
using `reduce_mean` instead of `reduce_sum` does  the division by `self.num_patches ** 2.0` for you